File: intelligence/ai_engine/discovery/relationship_mapper.py
"""Relationship mapping - builds entity-relationship graph from FK and inferred relationships."""

from collections import Counter
from typing import Any


class RelationshipMapper:
    """Maps explicit (FK) and inferred relationships between tables."""

    def map_relationships(
        self,
        schema: dict[str, Any],
        profiles: dict[str, dict],
    ) -> dict[str, Any]:
        relationships = []

        # 1. Explicit: foreign keys from schema
        for fk in schema.get("foreign_keys", []):
            relationships.append({
                "type": "explicit_fk",
                "source_table": fk["source_table"],
                "source_column": fk["source_column"],
                "target_table": fk["target_table"],
                "target_column": fk["target_column"],
                "confidence": 1.0,
            })

        # 2. Inferred: columns named "{table}_id" that match another table
        all_tables = {t["table_name"] for t in schema.get("tables", [])}
        explicit_pairs = {
            (r["source_table"], r["source_column"])
            for r in relationships
        }

        columns = schema.get("columns", {})
        for table_name, table_cols in columns.items():
            for col_info in table_cols:
                col_name = col_info["column_name"]
                if not col_name.endswith("_id"):
                    continue
                if (table_name, col_name) in explicit_pairs:
                    continue

                candidate = col_name[:-3]  # strip "_id"
                target_table = None
                if candidate in all_tables:
                    target_table = candidate
                elif f"{candidate}s" in all_tables:
                    target_table = f"{candidate}s"
                elif f"{candidate}es" in all_tables:
                    target_table = f"{candidate}es"

                if target_table and target_table != table_name:
                    relationships.append({
                        "type": "inferred",
                        "source_table": table_name,
                        "source_column": col_name,
                        "target_table": target_table,
                        "target_column": "id",
                        "confidence": 0.8,
                    })

        hub_entity = self._detect_hub_entity(relationships)
        entity_graph = self._build_adjacency_list(relationships)

        return {
            "relationships": relationships,
            "hub_entity": hub_entity,
            "entity_graph": entity_graph,
        }

    def _detect_hub_entity(self, relationships: list[dict]) -> str | None:
        """The hub entity is the table most referenced by foreign keys."""
        inbound = Counter()
        for rel in relationships:
            inbound[rel["target_table"]] += 1
        if not inbound:
            return None
        return inbound.most_common(1)[0][0]

    def _build_adjacency_list(self, relationships: list[dict]) -> dict[str, list[str]]:
        graph: dict[str, list[str]] = {}
        for rel in relationships:
            src = rel["source_table"]
            tgt = rel["target_table"]
            if src not in graph:
                graph[src] = []
            if tgt not in graph[src]:
                graph[src].append(tgt)
            if tgt not in graph:
                graph[tgt] = []
            if src not in graph[tgt]:
                graph[tgt].append(src)
        return graph

File: backend/src/intelligence/discovery/relationshipMapper.ts
import { SchemaSnapshot, ForeignKey, ColumnInfo } from './schemaInspector';

export interface Relationship {
  source_table: string;
  source_column: string;
  target_table: string;
  target_column: string;
  type: 'foreign_key' | 'inferred';
  confidence: number;
}

export interface RelationshipResult {
  relationships: Relationship[];
  hub_entity: string | null;
  entity_graph: Record<string, string[]>;
}

function buildFromForeignKeys(fks: ForeignKey[]): Relationship[] {
  return fks.map((fk) => ({
    source_table: fk.source_table,
    source_column: fk.source_column,
    target_table: fk.target_table,
    target_column: fk.target_column,
    type: 'foreign_key' as const,
    confidence: 1.0,
  }));
}

function inferFromNaming(
  columns: Record<string, ColumnInfo[]>,
  tableNames: Set<string>,
  existingPairs: Set<string>
): Relationship[] {
  const inferred: Relationship[] = [];

  for (const [tableName, cols] of Object.entries(columns)) {
    for (const col of cols) {
      const name = col.column_name.toLowerCase();
      if (!name.endsWith('_id') || name === 'id') continue;

      // Try to find the target table
      const prefix = name.slice(0, -3); // remove _id
      const candidates = [
        prefix + 's',        // user_id -> users
        prefix + 'es',       // process_id -> processes
        prefix,              // campaign_id -> campaign
        prefix.replace(/_/g, '') + 's', // admin_user_id -> adminusers
      ];

      for (const candidate of candidates) {
        if (tableNames.has(candidate) && candidate !== tableName) {
          const pairKey = `${tableName}.${col.column_name}->${candidate}`;
          if (!existingPairs.has(pairKey)) {
            inferred.push({
              source_table: tableName,
              source_column: col.column_name,
              target_table: candidate,
              target_column: 'id',
              type: 'inferred',
              confidence: 0.8,
            });
            existingPairs.add(pairKey);
          }
          break;
        }
      }
    }
  }

  return inferred;
}

export function detectHubEntity(relationships: Relationship[]): string | null {
  const refCounts: Record<string, number> = {};
  for (const rel of relationships) {
    refCounts[rel.target_table] = (refCounts[rel.target_table] || 0) + 1;
  }

  let maxTable: string | null = null;
  let maxCount = 0;
  for (const [table, count] of Object.entries(refCounts)) {
    if (count > maxCount) {
      maxCount = count;
      maxTable = table;
    }
  }

  return maxCount >= 2 ? maxTable : null;
}

function buildAdjacencyList(relationships: Relationship[]): Record<string, string[]> {
  const graph: Record<string, Set<string>> = {};
  for (const rel of relationships) {
    if (!graph[rel.source_table]) graph[rel.source_table] = new Set();
    if (!graph[rel.target_table]) graph[rel.target_table] = new Set();
    graph[rel.source_table].add(rel.target_table);
    graph[rel.target_table].add(rel.source_table);
  }

  const result: Record<string, string[]> = {};
  for (const [table, neighbors] of Object.entries(graph)) {
    result[table] = [...neighbors];
  }
  return result;
}

export function mapRelationships(snapshot: SchemaSnapshot): RelationshipResult {
  const tableNames = new Set(snapshot.tables.map((t) => t.table_name));

  // Explicit foreign keys
  const fkRelationships = buildFromForeignKeys(snapshot.foreign_keys);

  // Track existing pairs to avoid duplicates
  const existingPairs = new Set(
    fkRelationships.map((r) => `${r.source_table}.${r.source_column}->${r.target_table}`)
  );

  // Inferred from naming patterns
  const inferredRelationships = inferFromNaming(snapshot.columns, tableNames, existingPairs);

  const allRelationships = [...fkRelationships, ...inferredRelationships];

  return {
    relationships: allRelationships,
    hub_entity: detectHubEntity(allRelationships),
    entity_graph: buildAdjacencyList(allRelationships),
  };
}
